Unify argument validation into shared source class

heaths opened the pull request:
Fixes #7547. For now, this only updates EventHub, but the PR will be updated to update KeyVault in which, for another PR (#7563) included a copy. @schaabs did have some concerns about what this does to the stack, but it does help keep the exceptions consistent, i.e. `ArgumentNullException` for null and `ArgumentException` for empty strings, which is consistent with the Framework ([example](https://docs.microsoft.com/en-us/dotnet/api/system.io.filestream.-ctor?view=netcore-2.2#System_IO_FileStream__ctor_System_String_System_IO_FileMode_)).

heaths committed:
Unify argument validation into shared source class

Fixes #7547

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Core/Guard.cs:
```diff
@@ -3,87 +3,31 @@
             }
 
-            if (String.IsNullOrWhiteSpace(argumentValue))
+            if (string.IsNullOrWhiteSpace(argumentValue))
```
For internal consistency with other uses within Event Hubs, I'd prefer that we leave `String`, rather than changing case for these.

pakrym replied:
Our code style guide mandates `string`

jsquire replied:
News to me; last that I had spoken to Alex about it we didn't have a formalized style guide.   Can you throw me a pointer to it?

pakrym replied:
https://github.com/Azure/azure-sdk-for-net/pull/5528#issuecomment-478807105

jsquire replied:
Fair enough; that was about the time of the last conversation that was had.   Not coming from CoreFx and not being overly familiar with their standards, the nuance was missed in the push for hitting release deadlines.   That'll be sorted out in Dozie's work stream with the `.editorconfig` to make it obvious going forward.

For now, I appreciate and agree with Heath's suggestion to keep things internally consistent and let the formatting pass do everything at once.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Core/Guard.cs:
```diff
@@ -3,87 +3,31 @@
-        public static void ArgumentNotEmptyOrWhitespace(string argumentName,
-                                                        string argumentValue)
+        /// <param name="argumentName">The name of the argument being considered.</param>
+        /// <exception cref="ArgumentException">The argument is empty or contains only white-space.</exception>
```
For consistency with other types in Event Hubs, I'd prefer that we remove the `<exception>` lines, but if you feel strongly about them, please add a blank before and after to match the formatting for other doc comments.

heaths replied:
What would be a reason for not documenting known possible extensions, similar to what .NET and other APIs do?

heaths replied:
Not strongly, no, but it does help developers - especially with developer exceptions (e.g. parameter X can't be null). Using nullability attributes throughout our code would be nice, but even if/when we do if code calling into us doesn't during compilation it's still not checked, meaning we still have to do runtime validation. Methods deeper down the call stack could throw plenty of other exceptions. I guess I always saw documented exceptions in the Framework as "here's some exceptions you should expect and probably handle". Less important, I admit, for developer exceptions but the fact they are documented does help indicate that certain parameters cannot be null/empty/out of range/etc.

So I'll probably keep these since `Argument` has them but I promise I won't touch other parts of the code. ;)

jsquire replied:
Good enough for me.  :)

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Core/Guard.cs:
```diff
@@ -92,17 +36,15 @@ internal static class Guard
+            if (argumentValue != null && argumentValue.Length > maximumLength)
             {
-                throw new ArgumentOutOfRangeException(String.Format(CultureInfo.CurrentCulture, Resources.ArgumentStringTooLong, argumentName, maxmimumLength), argumentName);
+                throw new ArgumentOutOfRangeException(argumentName, $"The argument '{argumentName}' cannot exceed {maximumLength} characters.");
```
Is there a reason that these are being changed from the current resource set?

heaths replied:
There are a few reasons.

1. Since we're using shared source for the main `Argument` class from Azure.Core, sharing the resources to support it would be extra work and error prone. And since we aren't for that, it would be inconsistent for some developer exceptions to be localized and others not.
2. For *developer exceptions* it's often easier to find answers when the exception isn't localized, especially when those exception cases are documented. ;) Talked with @pakrym who brought this up with cases he's seen.
3. No other track 2 service is currently localizing developer exceptions.

jsquire replied:
To be transparent, the only reason that the strings are resources was really inertia from track one.  They're not truly localized, only lightly applying culture formatting to dates and numeric values.   

I find value in having a centralized dictionary of messages just to keep verbiage consistent across the code base, but as `Argument` is already a central point where messages aren't re-used, I've no objections to keeping the changes.   I asked more from curiosity than any objection.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/tests/Authorization/SharedAccessSignatureTests.cs:
```diff
@@ -56,7 +57,9 @@ public void ToStringReflectsTheValue()
         public void CompositeConstructorValidatesTheResource(string resource)
         {
-            Assert.That(() => new SharedAccessSignature(resource, "Yay", "OMG!", TimeSpan.FromMilliseconds(500)), Throws.ArgumentException);
+            ExactTypeConstraint typeConstraint = resource is null ? Throws.ArgumentNullException : Throws.ArgumentException;
```
*Note to self: *  These feel like a good candidate for an extension method.

heaths replied:
If you're thinking of extending what `That()` returns, I agree. xUnit - with which I'm more familiar - also has `Assert.Throws<T>(string paramName, Action action) where T : ArgumentException` that allow you to validate a particular parameter name (i.e. make sure you're getting an `ArgumentException` for the right parameter). Such an extension method might also do that.

jsquire replied:
I was thinking more of doing a string extension within the tests, since we're generating that constraint frequently.   Something like:

```csharp
Assert.That() => new Blah(stringValue, stringValue.GetExceptionConstraint());
```

Or something less awkward.   I haven't fully thought it through, but may play with it as time allows.

jsquire replied:
If you weren't concerned with precision, you could have done these as:

```csharp
Assert.That(() => new Blah(stringValue), Throws.InstanceOf<ArgumentException>());
```

You wouldn't be proving ArgumentNull vs Argument, but it's probably good enough outside the `Argument` tests themselves.

jsquire reviewed (changes requested):
LGTM.    I asked for a couple of small changes, mostly around formatting, for internal consistency within the Event Hubs library.  

Event Hubs has a release related to preview 3 due this week, and I'd like to request that this not be merged until we've passed that milestone.  I'm tagging "Request Changes" only to prevent an unintended merge beforehand; I don't see any blockers otherwise.

heaths committed:
Resolve PR feedback

heaths committed:
Add nullability attributes to Argument helper

Will help prevent "possible null" errors later when we enable nullability for code that uses this class.

heaths committed:
Fix breaks with opt-in nullable attributes

heaths committed:
Unify argument validation across .NET clients

Fixes #7547

heaths committed:
Fix Azure.Core.Tests

Forgot to update this project in another solution.

heaths committed:
Don't unnecessarily include nullable attributes

jsquire reviewed (approved):
Event Hubs should be clear.   Thanks for your patience.

heaths committed:
Fix issue from rebase
